sycl/include/sycl/ext/intel/experimental/esimd/detail/math_intrin.hpp

```diff
@@ -946,127 +945,57 @@ __esimd_ufbh(__SEIEED::vector_type_t<T0, SZ> src0) {
   }
 
   return retv;
-};
-
-template <int SZ>
-__ESIMD_INTRIN __SEIEED::vector_type_t<float, SZ>
-__esimd_inv(__SEIEED::vector_type_t<float, SZ> src0) {
-  __SEIEED::vector_type_t<float, SZ> retv;
-
-  for (int i = 0; i < SZ; i++) {
-    SIMDCF_ELEMENT_SKIP(i);
-    retv[i] = 1.f / src0[i];
-  }
-  return retv;
-};
-
-template <int SZ>
-__ESIMD_INTRIN __SEIEED::vector_type_t<float, SZ>
-__esimd_log(__SEIEED::vector_type_t<float, SZ> src0) {
-  __SEIEED::vector_type_t<float, SZ> retv;
+}
 
-  for (int i = 0; i < SZ; i++) {
-    SIMDCF_ELEMENT_SKIP(i);
-    retv[i] = logf(src0[i]) / logf(2.);
+// Host intrinsics are implemented via converting elements to enclosing Cpp
+// type (always 'float' except ieee_sqrt, which can be 'double'), applying
+// standard C++ library math function and converting back to the element type.
+//
+#define __ESIMD_UNARY_EXT_MATH_HOST_INTRIN(name, formula)                      \
+  template <class T, int SZ>                                                   \
+  __ESIMD_INTRIN __ESIMD_raw_vec_t(T, SZ)                                      \
+      __esimd_##name(__ESIMD_raw_vec_t(T, SZ) src) {                           \
+    using CppT = __SEIEED::__cpp_t<T>;                                         \
+    using CppVecT = __ESIMD_cpp_vec_t(T, SZ);                                  \
+    CppVecT ret_cpp{0};                                                        \
+    CppVecT src_cpp = __SEIEED::convert_vector<CppT, T, SZ>(src);              \
+                                                                               \
+    for (int i = 0; i < SZ; i++) {                                             \
+      SIMDCF_ELEMENT_SKIP(i);                                                  \
+      ret_cpp[i] = formula;                                                    \
+    }                                                                          \
+    __ESIMD_raw_vec_t(T, SZ) ret =                                             \
+        __SEIEED::convert_vector<T, CppT, SZ>(ret_cpp);                        \
+    return ret;                                                                \
   }
-  return retv;
-};
-template <int SZ>
-__ESIMD_INTRIN __SEIEED::vector_type_t<float, SZ>
-__esimd_exp(__SEIEED::vector_type_t<float, SZ> src0) {
-  __SEIEED::vector_type_t<float, SZ> retv;
 
-  for (int i = 0; i < SZ; i++) {
-    SIMDCF_ELEMENT_SKIP(i);
-    retv[i] = powf(2.f, src0[i]);
-  }
-  return retv;
-};
-template <int SZ>
-__ESIMD_INTRIN __SEIEED::vector_type_t<float, SZ>
-__esimd_sqrt(__SEIEED::vector_type_t<float, SZ> src0) {
-  __SEIEED::vector_type_t<float, SZ> retv;
+__ESIMD_UNARY_EXT_MATH_HOST_INTRIN(inv, 1.f / src_cpp[i])
+__ESIMD_UNARY_EXT_MATH_HOST_INTRIN(log, logf(src_cpp[i]) / logf(2.f))
+__ESIMD_UNARY_EXT_MATH_HOST_INTRIN(exp, powf(2.f, src_cpp[i]))
```

Comment:
In all cases other than ieee_sqrt this will lead to increased accuracy of host calculations (double) compared to GPU (half, float), so yes, that was intentional, as seemув good enough for the test.
anyway, will try your suggestion.

Reply:
doesn't std::log(src_cpp[i]) call float32 version because src_cpp[i] is 'float'.
See https://en.cppreference.com/w/cpp/numeric/math/log
"float       log ( float arg );"

Reply:
yes, but std::log(2.0) will return double and call double version of log.

Reply:
... so std::log(src_cpp[i]) / std::log(2.0)) will return double

Reply:
let's actually leave the host implementation as is here - as I mentioned, none of these can be double except ieee_sqrt (which is handled correctly - will use double for calculations), but we don't want host to perform calculations with unnecessarily increased accuracy and use double for log and pow.